packages/service-library/src/servicelib/docker_utils.py

```diff
@@ -12,3 +28,188 @@ def to_datetime(docker_timestamp: str) -> datetime:
+    # NOTE: 1 subprogress per layer, then subprogress of size 2 with weights?
+    # NOTE: or, we compute the total size X2, then upgrade that based on the status? maybe simpler
+
+    @dataclass
+    class PulledStatus:
+        size: int
+        downloaded: int = 0
+        extracted: int = 0
+
+    layer_id_to_size = {
+        layer.digest.removeprefix("sha256:")[:12]: PulledStatus(layer.size)
+        for layer in image_information.layers
+    }
+    image_layers_total_size = sum(layer.size for layer in image_information.layers) * 2
+    image_name = image.split("/")[-1]
+    async with (
+        aiodocker.Docker() as client,
+        progress_bar.sub_progress(image_layers_total_size) as sub_progress,
+    ):
+        async for pull_progress in client.images.pull(
+            image, stream=True, auth=registry_auth
+        ):
+            parsed_progress = parse_obj_as(_DockerPullImage, pull_progress)
+            match parsed_progress.status.lower():
+                case progress_status if any(
+                    msg in progress_status
+                    for msg in [
+                        "pulling from",
+                        "pulling fs layer",
+                        "waiting",
+                        "digest: ",
+                        "status: downloaded newer image for ",
+                        "status: image is up to date for ",
+                        "already exists",
+                    ]
+                ):
+                    # nothing to do here
+                    pass
+                case "downloading":
+                    assert parsed_progress.id  # nosec
+                    assert parsed_progress.progress_detail  # nosec
+                    assert parsed_progress.progress_detail.current  # nosec
+                    layer_id_to_size[
+                        parsed_progress.id
+                    ].downloaded = parsed_progress.progress_detail.current
+                case "verifying checksum" | "download complete":
+                    assert parsed_progress.id  # nosec
+                    layer_id_to_size[parsed_progress.id].downloaded = layer_id_to_size[
+                        parsed_progress.id
+                    ].size
+                case "extracting":
+                    assert parsed_progress.id  # nosec
+                    assert parsed_progress.progress_detail  # nosec
+                    assert parsed_progress.progress_detail.current  # nosec
+                    layer_id_to_size[
+                        parsed_progress.id
+                    ].extracted = parsed_progress.progress_detail.current
+                case "pull complete":
+                    assert parsed_progress.id  # nosec
+                    layer_id_to_size[parsed_progress.id].extracted = layer_id_to_size[
+                        parsed_progress.id
+                    ].size
+                case _:
```

Comment:
So do I understand correctly that this fixes the progress bar for the pulling of the docker images (meaning the data which is part of the service, as opposed to the user data)?

Reply:
With these changes does the progressbar progress smoothly? Because if that's the case we could give a ETA. I think that's one of the points the POs were pushing for in the last review meetings.

Reply:
Cool, thanks for the clarification. Yes, I think it would be great if we did some statistics so that we could maybe create a "smooth" progress bar together with a time estimate at some point